### DISCUSSION TITLE
Proper way to create a custom node abstracting ruby annotation

### DISCUSSION THREAD

**Discussion by e-e** [facebook/lexical#3666] [Q&A]
I would like to create a custom node to abstract [ruby annotations](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ruby), and as I'm fairly new to Lexical, I'm struggling to conceptualize how that might be achieved. I had originally thought I could use a single node extending `ElementNode`, but it seems that the `createDOM` method is only intended to return an empty element of the desired type? Do I need to create a custom node for each of the child elements involved in the annotation? If someone could point me in the right direction, that would be appreciated. Thank you in advance!

**Comment by moy2010** [comment]
Without more information, it's hard to guess what/how are you trying to achieve. For instance, I don't understant what you meant on the `createDOM` part.

I would start by pointing you to the [documentation for nodes](https://lexical.dev/docs/concepts/nodes), So that you get a grasp on the difference between element nodes and text nodes, for example.

**Reply by e-e** [reply]
I'll try to provide some more context by way of an example. Taking the example from the above linked documentation for the `<ruby>` tag, let's say I want to annotate the characters "漢字" when they are typed into the editor. The html for that would look like:
```
<ruby>
      漢 <rp>(</rp><rt>Kan</rt><rp>)</rp>
      字 <rp>(</rp><rt>ji</rt><rp>)</rp>
</ruby>
```
and would be rendered like:
<img width="771" alt="Screen Shot 2023-01-04 at 11 42 06 AM" src="https://user-images.githubusercontent.com/5191838/210474692-7ac6fb93-1a0d-4568-a66c-24fd27566244.png">

I had taken some time to read the documentation on nodes prior to posting my question, and the example that is given for extending `ElementNode` defines a `createDOM` method that creates an empty `p` element and returns it without any content:

```
createDOM(): HTMLElement {
    // Define the DOM element here
    const dom = document.createElement('p');
    return dom;
  }
```

This makes sense to me for a node type that would be use as a container for additional input text, and it is unclear if that is the sole intended use for the method. But in my case I am assuming that I will already have all of the info required for the new node (the characters to annotate and the annotations themselves). 

This is what I currently have, which kind of works, but I'm also getting `TypeError: selectionTarget.getBoundingClientRect is not a function`:

```typescript
export default class AnnotationNode extends ElementNode {
  __character: string;
  __reading: string;

  constructor(character: string, reading: string, key?: NodeKey) {
    super(key);

    this.__character = character;
    this.__reading = reading;
  }

  static getType(): string {
    return "annotation-node";
  }

  getCharacter(): string {
    const self = this.getLatest();
    return self.__character;
  }

  getReading(): string {
    const self = this.getLatest();
    return self.__reading;
  }

  /**
   * example:
   * <ruby>
   *   漢 <rp>(</rp><rt>Kan</rt><rp>)</rp>
   *   字 <rp>(</rp><rt>ji</rt><rp>)</rp>
   * </ruby>
   */
  createDOM(): HTMLElement {
    const dom = document.createElement("ruby");
    dom.innerHTML = `${this.getCharacter()}<rp>(</rp><rt>${this.getReading()}</rt><rp>)</rp>`;
    return dom;
  }

  updateDOM(prevNode: AnnotationNode, dom: HTMLElement): boolean {
    // Returning false tells Lexical that this node does not need its
    // DOM element replacing with a new copy from createDOM.
    return false;
  }

  static clone(node: AnnotationNode): AnnotationNode {
    return new AnnotationNode(
      node.getCharacter(),
      node.getReading(),
      node.__key
    );
  }
}
```

The above is a simplified example, but what I'm looking to eventually be able to do is listen for when a sentence ending punctuation character is entered, make an API request with that sentence text, retrieve the annotations, then replace the original text with the annotated version. I imagine I would want an additional node type to wrap an annotated sentence (SentenceNode?) in order to make it easy to filter out sentences that are already annotated. This node would then be made up of simple text nodes, for the parts of the sentence that don't have any annotations, and the annotated nodes.

I guess my question is 1) what am I missing that is causing the getBoundingClientRect error? And 2) does this approach make any sense with respect to Lexical conventions?

**Reply by chalecki** [reply]
> This is what I currently have, which kind of works, but I'm also getting `TypeError: selectionTarget.getBoundingClientRect is not a function:

Could you show more detail about this Error and where it comes from? 
Are you using Lexical in React component or plain JS?

**Reply by e-e** [reply]
Using React, yes.

Here is more of the error:
```
×
TypeError
selectionTarget.getBoundingClientRect is not a function
updateDOMSelection
https://c6j3xg.csb.app/node_modules/lexical/Lexical.dev.js:5824:45
commitPendingUpdates
https://c6j3xg.csb.app/node_modules/lexical/Lexical.dev.js:6271:9
eval
https://c6j3xg.csb.app/node_modules/lexical/Lexical.dev.js:6594:9
This screen is visible only in development. It will not appear if the app crashes in production.
Open your browser’s developer console to further inspect this error.
This error overlay is powered by `react-error-overlay` used in `create-react-app`.
```

**Comment by jmy2022** [comment]
How the end
